lib/api/core/vault.js

```diff
@@ -0,0 +1,218 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-console */
+
+'use strict';
+
+const
+  crypto = require('crypto'),
+  _ = require('lodash'),
+  fs = require('fs');
+
+const defaultSecretsFile = '../../../config/secrets.enc.json';
+
+class Vault {
+  static generateSeed () {
+    return crypto.randomBytes(8).toString('hex');
+  }
+
+  constructor (config) {
+    this.config = config;
+    this.secrets = {};
+  }
+
+  /**
+   * Prepare crypto primitives.
+   * Use the key passed in parameter or the KUZZLE_VAULT_KEY
+   * environment variable.
+   *
+   * @param {string|null} vaultKey - key used to decrypt the secrets
+   */
+  prepareCrypto (vaultKey = null) {
+    if (_.isString(vaultKey) && vaultKey.length > 0) {
+      this.vaultKey = vaultKey;
+    }
+    else if (_.isString(process.env.KUZZLE_VAULT_KEY)
+      && process.env.KUZZLE_VAULT_KEY.length > 0
+    ) {
+      this.vaultKey = process.env.KUZZLE_VAULT_KEY;
+    }
+    else {
+      return;
+    }
+
+    if (this.config.server.vaultSeed === '764c1afa2988981a') {
+      console.warn(
+        'You are using the default vault seed.\n',
+        'It\'s STRONGLY advised to change this value.\n',
+        'Run the following command to generate a new seed:\n',
+        '\tbin/kuzzle vaultSeed\n'
+      );
+    }
+
+    this.vaultKeyHash = crypto.createHash('md5')
+      .update(this.vaultKey, 'utf-8')
+      .digest('hex')
+      .toUpperCase();
+
+    this.cipherIV = this.config.server.vaultSeed.substr(0, 16);
+  }
+
+  /**
+   * Init the vault with a decryption key and a file containing
+   * the secrets, decrypt it and store decrypted secrets
+   *
+   * Rejects with an error if:
+   *  - a key is found but no file
+   *  - a file is found but no key
+   *
+   * @param {string|null} secretsFile - file containing the encrypted secrets
+   * @returns {Promise}
+   */
+  init (secretsFile = null) {
+    this.encryptedSecretsFile = secretsFile || defaultSecretsFile;
+
+    const fileExists = fs.existsSync(this.encryptedSecretsFile);
```

Comment:
Do you check permissions on the file somewhere ?

Reply:
Node.js will throw a `[Errno 13] Permission denied` if the permissions are not sufficient

Reply:
oh ok I didn't know about Errno in js

Reply:
Actually it's still the system that throw this error, Node.js just forward it